### ISSUE TITLE
Avoid data duplication

### ISSUE THREAD

**Issue by authsec** [CyberReboot/CRviz#355]
### Describe the feature you are requesting
I would like a mechanism that avoids data duplication where I can reference a data object from with the JSON structure.

### Describe your use case

I want to visualize firewall rulesets using CRviz. The rules are applied/requested through a change request. Now the exact same change request needs to be attached to each server, say s1,s2,s3,s4 or a server group, say frontend (s1,s2), backend(s3,s4).

To avoid a structure like:

```
s1 {
ip: ...
crq: {
id: 123
type: firewall
from: 1.2.3.4
port: 443
}
}

s2 {
ip: ...
crq: {
id: 123
type: firewall
from: 1.2.3.4
port: 443
}
}
```

I'd like to be able to do s.th. like

```
crq: {
id: 0001
type: firewall
from: 1.2.3.4
port: 443
}

s1 {
ip: ...
crq-id: 0001 
}

s2 {
ip: ...
crq-id: 0001
}
```

### Explain how this feature supports your use case? Other use cases?

As I am managing the JSON information manually at the moment, this would help me avoid duplicating data and support the DRY principle.

This feature might reduce the data volume required to be transferred/processed/stored significantly depending on the use case.

### Additional Info

If this is already possible, I would find it helpful if you could include an example how to implement this.

**Comment by rashley-iqt** [comment]
Thank you for the feedback. I really like the idea but I need to think through how to go about implementing something like this. If you have an example of the kind of file you would like to import that you can share that would be incredibly helpful.

**Comment by authsec** [comment]
Thank you for considering this!

I created an exemplary structure which I hope helps illustrates the basic idea and works as a starting point. In this example Ideally I'd be able to use CRQs as a generic grouping class, as e.g. "Environment" would be.

[crq-sample.txt](https://github.com/CyberReboot/CRviz/files/3258941/crq-sample.txt)

**Comment by rashley-iqt** [comment]
I think I have a solution for this. [JSON RFC 6901](https://tools.ietf.org/html/rfc6901) defines a way for a JSON document to reference either its own sub elements, or elements of a remote JSON document. It will require some changes to CRviz that I have a proof of concept for but still need to work out some minor implementation details of and then test. It would also require some slight tweaks to the file structure you had initially provided. I have attached examples below of what the file structures would look like.

essentially an object is referenced using an XPath like structure using the `ref$` key. Like this:
`{"property": {"ref$":"#/path/to/value"}}`
In the case that the referenced object is part of an array, the desired item is referenced by the path to the array and then the ordinal: `{"property": {"ref$":"#/path/to/array/index"}}` This is demonstrated in the `_array` example. If you prefer to use a more readable key then you would want your values to be part of a dictionary and use a path of the form `{"property": {"ref$":"#/path/to/dict/key"}}`. you can see this form in the `_dict` example. Personally, I prefer the dictionary form because i think it is easier to read, particularly in large files but the solution i am working toward would work with either.

@authsec Would this get you to the functionality you are looking for?

[crq-sample_array.txt](https://github.com/CyberReboot/CRviz/files/3308417/crq-sample_array.txt)
[crq-sample_dict.txt](https://github.com/CyberReboot/CRviz/files/3308418/crq-sample_dict.txt)

**Comment by authsec** [comment]
This is awesome! That would definitely give me the functionality I am looking for. I prefer the dictionary approach over the array as well, it seems way better to read.

From a visualization point of view, can this be implemented in such a way that you can use the dictionary key as a display name? So when you drill down grouping `Environment` then `IP` you'd see a circle `CRQ00002345` and a circle `CRQ00003456` or maybe just a circle `CRQs`? I guess I'm generally talking about grouping on array data.

*Side note*: When navigating the dataset, with the mentioned grouping it might be necessary to display the other information as leaf nodes. Maybe you can use the leaf nodes to start navigating the dataset again from there. So if I were to hit `Environment` node it would start with the `Environment` overview (Development, Testing, Production) again.

I created a little drawing roughly illustrating the idea.

![crq-viz](https://user-images.githubusercontent.com/7973409/59975510-c0a3eb80-95b8-11e9-9663-642f20c15335.png)

**Comment by rashley-iqt** [comment]
To kind of close the loop on this one you are absolutely right, @authsec. We could do a much better job with how we visualize array content however, I want to keep this issue scoped to adding the refs to the data. If you want to open an issue for the array content visualization that would be great. Our whole team really appreciates you taking the time to get involved so please keep the suggestions coming.

**Comment by rashley-iqt** [comment]
As a final note @authsec, this feature was released in version 0.3.5, so please test it out and let me know how it works for you.
